### ISSUE TITLE
gcext segfaults

### ISSUE THREAD

**Issue by staticfloat** [JuliaLang/julia#38504]
Running the `tests/gcext` tests segfaults on the latest `master` (72e67d707ba1f399f082e149080a5eaa24bc63a0):

```
$ ./gcext 
MethodError: no method matching find_package(Internal error: encountered unexpected error in runtime:
ReadOnlyMemoryError()
task_scanner at ./gcext (unknown line)
gc_mark_loop at /home/sabae/src/julia-master/src/gc.c:2585
_jl_gc_collect at /home/sabae/src/julia-master/src/gc.c:2973
jl_gc_collect at /home/sabae/src/julia-master/src/gc.c:3179
maybe_collect at /home/sabae/src/julia-master/src/gc.c:828 [inlined]
jl_gc_pool_alloc at /home/sabae/src/julia-master/src/gc.c:1143
jl_gc_alloc_ at /home/sabae/src/julia-master/src/julia_internal.h:282 [inlined]
jl_gc_alloc at /home/sabae/src/julia-master/src/gc.c:3221
_new_array_ at /home/sabae/src/julia-master/src/array.c:122 [inlined]
_new_array at /home/sabae/src/julia-master/src/array.c:190 [inlined]
jl_alloc_array_1d at /home/sabae/src/julia-master/src/array.c:461
Array at ./boot.jl:448 [inlined]
Array at ./boot.jl:467 [inlined]
getindex at ./array.jl:391 [inlined]
simple_dce! at ./compiler/ssair/ir.jl:1358
getfield_elim_pass! at ./compiler/ssair/passes.jl:724
run_passes at ./compiler/ssair/driver.jl:132
optimize at ./compiler/optimize.jl:186
_typeinf at ./compiler/typeinfer.jl:239
typeinf at ./compiler/typeinfer.jl:211
abstract_call_method_with_const_args at ./compiler/abstractinterpretation.jl:307
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:154
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
abstract_call_method_with_const_args at ./compiler/abstractinterpretation.jl:307
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:154
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_edge at ./compiler/typeinfer.jl:745
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_edge at ./compiler/typeinfer.jl:745
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_edge at ./compiler/typeinfer.jl:745
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_edge at ./compiler/typeinfer.jl:745
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_ext at ./compiler/typeinfer.jl:835
typeinf_ext_toplevel at ./compiler/typeinfer.jl:868
typeinf_ext_toplevel at ./compiler/typeinfer.jl:864
jfptr_typeinf_ext_toplevel_12333 at /home/sabae/src/julia-master/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /home/sabae/src/julia-master/src/gf.c:2191 [inlined]
jl_apply_generic at /home/sabae/src/julia-master/src/gf.c:2373
jl_apply at /home/sabae/src/julia-master/src/julia.h:1693 [inlined]
jl_type_infer at /home/sabae/src/julia-master/src/gf.c:300
jl_generate_fptr at /home/sabae/src/julia-master/src/jitlayers.cpp:316
jl_compile_method_internal at /home/sabae/src/julia-master/src/gf.c:1924
jl_compile_method_internal at /home/sabae/src/julia-master/src/gf.c:1878 [inlined]
_jl_invoke at /home/sabae/src/julia-master/src/gf.c:2183 [inlined]
jl_apply_generic at /home/sabae/src/julia-master/src/gf.c:2373
showerror at ./errorshow.jl:265
unknown function (ip: 0x7f3fa7f6af48)
_jl_invoke at /home/sabae/src/julia-master/src/gf.c:2191 [inlined]
jl_apply_generic at /home/sabae/src/julia-master/src/gf.c:2373
jl_apply at /home/sabae/src/julia-master/src/julia.h:1693 [inlined]
jl_call2 at /home/sabae/src/julia-master/src/jlapi.c:247
checked_eval_string at ./gcext (unknown line)
main at ./gcext (unknown line)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at ./gcext (unknown line)
::Module, Internal error: encountered unexpected error in runtime:
OutOfMemoryError()
jl_gc_big_alloc at /home/sabae/src/julia-master/src/gc.c:894
jl_gc_alloc_ at /home/sabae/src/julia-master/src/julia_internal.h:287 [inlined]
jl_gc_alloc at /home/sabae/src/julia-master/src/gc.c:3221
jl_alloc_svec_uninit at /home/sabae/src/julia-master/src/simplevector.c:60
jl_specializations_get_linfo at /home/sabae/src/julia-master/src/gf.c:160
specialize_method at ./compiler/utilities.jl:144 [inlined]
specialize_method at ./compiler/utilities.jl:134 [inlined]
typeinf_edge at ./compiler/typeinfer.jl:708
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_edge at ./compiler/typeinfer.jl:745
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_edge at ./compiler/typeinfer.jl:745
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_edge at ./compiler/typeinfer.jl:745
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_edge at ./compiler/typeinfer.jl:745
abstract_call_method at ./compiler/abstractinterpretation.jl:466
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:130
abstract_call_known at ./compiler/abstractinterpretation.jl:1003
abstract_call at ./compiler/abstractinterpretation.jl:1026
abstract_call at ./compiler/abstractinterpretation.jl:1010
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1132
typeinf_local at ./compiler/abstractinterpretation.jl:1386
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1442
_typeinf at ./compiler/typeinfer.jl:216
typeinf at ./compiler/typeinfer.jl:211
typeinf_ext at ./compiler/typeinfer.jl:835
typeinf_ext_toplevel at ./compiler/typeinfer.jl:868
typeinf_ext_toplevel at ./compiler/typeinfer.jl:864
jfptr_typeinf_ext_toplevel_12333 at /home/sabae/src/julia-master/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /home/sabae/src/julia-master/src/gf.c:2191 [inlined]
jl_apply_generic at /home/sabae/src/julia-master/src/gf.c:2373
jl_apply at /home/sabae/src/julia-master/src/julia.h:1693 [inlined]
jl_type_infer at /home/sabae/src/julia-master/src/gf.c:300
jl_generate_fptr at /home/sabae/src/julia-master/src/jitlayers.cpp:316
jl_compile_method_internal at /home/sabae/src/julia-master/src/gf.c:1924
jl_compile_method_internal at /home/sabae/src/julia-master/src/gf.c:1878 [inlined]
_jl_invoke at /home/sabae/src/julia-master/src/gf.c:2183 [inlined]
jl_apply_generic at /home/sabae/src/julia-master/src/gf.c:2373
showerror at ./errorshow.jl:265
unknown function (ip: 0x7f3fa7f6af48)
_jl_invoke at /home/sabae/src/julia-master/src/gf.c:2191 [inlined]
jl_apply_generic at /home/sabae/src/julia-master/src/gf.c:2373
jl_apply at /home/sabae/src/julia-master/src/julia.h:1693 [inlined]
jl_call2 at /home/sabae/src/julia-master/src/jlapi.c:247
checked_eval_string at ./gcext (unknown line)
main at ./gcext (unknown line)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at ./gcext (unknown line)
::String)
Closest candidates are:
  find_package(::Any) at loading.jl:239
```

Pinging @fingolfin as I believe you are the last person to touch this file.  :)

**Comment by fingolfin** [comment]
See issue #38479 -- @rbehrends is already working on this.

**Comment by fingolfin** [comment]
Also, it would be great if gcext (and perhaps also the `embedding` test?) could be run regularly as part of the Julia CI, then we might notice these issues earlier? Or perhaps it already is, and we / I just don't know where to look for it?

**Comment by staticfloat** [comment]
It is not, I agree it should be.

**Comment by rbehrends** [comment]
I think this should be fixed in #38510. It's the result of some code that was (unnecessarily) copied from the embedding test and not updated when the underlying Julia API for `find_package()` changed.

As for running this as part of the regular tests, I'd have to know how to properly integrate it into the test suite. This would require updating `test/choosetests.jl`, I think, but I'm not sure how to hook up a test that doesn't run via the normal Julia executable. (The same goes for the embedding test.)
